src/Decred/Signer.cpp

```diff
@@ -0,0 +1,199 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Signer.h"
+
+#include "TransactionInput.h"
+#include "TransactionOutput.h"
+
+#include "../BinaryCoding.h"
+#include "../Hash.h"
+#include "../HexCoding.h"
+
+#include <TrustWalletCore/TWBitcoinOpCodes.h>
+
+using namespace TW;
+using namespace TW::Decred;
+
+Result<Transaction> Signer::sign() {
+    signedInputs.clear();
+    std::copy(std::begin(transaction.inputs), std::end(transaction.inputs), std::back_inserter(signedInputs));
+
+    const bool hashSingle = ((input.hash_type() & ~TWSignatureHashTypeAnyoneCanPay) == TWSignatureHashTypeSingle);
+    for (auto i = 0; i < plan.utxos.size(); i += 1) {
+        auto& utxo = plan.utxos[i];
+
+        // Only sign TWSignatureHashTypeSingle if there's a corresponding output
+        if (hashSingle && i >= transaction.outputs.size()) {
+            continue;
+        }
+        auto script = Bitcoin::Script(utxo.script().begin(), utxo.script().end());
+        auto result = sign(script, i);
+        if (!result) {
+            return Result<Transaction>::failure(result.error());
+        }
+        signedInputs[i].script = result.payload();
+    }
+
+    Transaction tx(transaction);
+    tx.inputs = move(signedInputs);
+    tx.outputs = transaction.outputs;
+    return Result<Transaction>::success(std::move(tx));
+}
+
+Result<Bitcoin::Script> Signer::sign(Bitcoin::Script script, size_t index) {
+    Bitcoin::Script redeemScript;
+    std::vector<Data> results;
+
+    auto result = signStep(script, index);
+    if (result) {
+        results = result.payload();
+    } else {
+        return Result<Bitcoin::Script>::failure(result.error());
+    }
+    auto txin = transaction.inputs[index];
+
+    if (script.isPayToScriptHash()) {
+        script = Bitcoin::Script(results.front().begin(), results.front().end());
+        auto result = signStep(script, index);
+        if (!result) {
+            return Result<Bitcoin::Script>::failure(result.error());
+        }
+        results = result.payload();
+        results.push_back(script.bytes);
+        redeemScript = script;
+        results.push_back(redeemScript.bytes);
+    }
+
+    return Result<Bitcoin::Script>::success(Bitcoin::Script(pushAll(results)));
+}
+
+Result<std::vector<Data>> Signer::signStep(Bitcoin::Script script, size_t index) {
```

Comment:
This part is similar to bitcoin, could we reuse it?

Reply:
Again, the types it uses are different. Also serialization format is different for the preimage and doesn't support segwit.